zipkin-redis/src/main/scala/com/twitter/zipkin/storage/redis/RedisIndex.scala

```diff
@@ -31,173 +32,157 @@ trait RedisIndex extends Index {
+    override def encodeKey(key: AnnotationKey) =
+      copiedBuffer("annotations:%s:%s".format(key.service, key.annotation), UTF_8)
+  }
+  private[this] lazy val binaryAnnotationIndex = new TraceIndex[BinaryAnnotationKey](database, ttl) {
+    override def encodeKey(key: BinaryAnnotationKey) =
+      copiedBuffer("binary_annotations:%s:%s:%s".format(key.service, key.annotation, key.value), UTF_8)
+  }
+  private[this] lazy val spanMap = new RedisSetMap(database, "span", ttl)
+  private[this] lazy val serviceArray = new RedisSet(database, "services")
+  // TODO: bad and misleading name. should be spanDurations?
+  private[this] lazy val spanDurations = new SpanDurationStore(database, "ttlMap")
 
   override def close() = {
     database.release()
   }
 
-  private[this] def zRangeResultsToSeqIds(arr: ZRangeResults): Seq[IndexedTraceId] =
-    arr.asTuples map (tup => IndexedTraceId(tup._1, tup._2.toLong))
-
-  private[redis] def redisJoin(items: String*) = items.mkString(":")
-
-  override def getTraceIdsByName(serviceName: String, span: Option[String],
-    endTs: Long, limit: Int): Future[Seq[IndexedTraceId]] =
-    serviceSpanMap.get(
-      serviceName,
-      span,
-      ttl map (dur => endTs - dur.inMicroseconds),
-      endTs,
-      limit) map zRangeResultsToSeqIds
-
-  override def getTraceIdsByAnnotation(serviceName: String, annotation: String, value: Option[ByteBuffer],
-    endTs: Long, limit: Int): Future[Seq[IndexedTraceId]] = (value match {
-      case Some(anno) =>
-        binaryAnnotationsListMap.get(
-          redisJoin(serviceName, annotation, ChannelBuffers.copiedBuffer(anno)),
-          (ttl map (dur => (endTs - dur.inMicroseconds).toDouble)).getOrElse(0.0),
-          endTs,
-          limit)
-      case None =>
-        annotationsListMap.get(
-          redisJoin(serviceName, annotation),
-          (ttl map (dur => (endTs - dur.inMicroseconds).toDouble)).getOrElse(0.0),
-          endTs,
-          limit)
-    }) map zRangeResultsToSeqIds
+  override def getTraceIdsByName(
+    serviceName: String, spanName: Option[String],
+    endTs: Long, limit: Int): Future[Seq[IndexedTraceId]] = {
+    if (spanName.isDefined) {
+      spanIndex.list(SpanKey(serviceName, spanName.get), endTs, limit)
+    } else {
+      serviceIndex.list(serviceName, endTs, limit)
+    }
+  }
+
+  override def getTraceIdsByAnnotation(
+    serviceName: String, annotation: String, value: Option[ByteBuffer],
+    endTs: Long, limit: Int): Future[Seq[IndexedTraceId]] = {
+    if (value.isDefined) {
+      binaryAnnotationIndex.list(BinaryAnnotationKey(serviceName, annotation, copiedBuffer(value.get).toString(UTF_8)), endTs, limit)
+    } else {
+      annotationIndex.list(AnnotationKey(serviceName, annotation), endTs, limit)
+    }
+  }
 
   override def getTracesDuration(traceIds: Seq[Long]): Future[Seq[TraceIdDuration]] = Future.collect(
     traceIds map (getTraceDuration(_))
-  ) map (_ flatten)
+  ) map (_.flatten)
```

Comment:
little things like this and a bunch of others means there are no more compiler warnings on this project